wgpu/gui/auto/qt.py

```diff
@@ -0,0 +1,270 @@
+from collections import defaultdict
+import importlib
+import sys
+import traceback
+
+for libname in ("PySide6", "PyQt6", "PySide2", "PyQt5"):
+    try:
+        importlib.import_module(libname)
+        break
+    except ModuleNotFoundError:
+        pass
+
+from ..qt import WgpuCanvas, QtCore, QtWidgets  # noqa: E402
+
+# Global reference to an app. Will be either instantiated or set to the
+# global Qt instance when the first QAutoWgpuCanvas object is created
+app = None
+
+
+def run():
+    app.exec() if hasattr(app, "exec") else app.exec_()
+
+
+def call_later(delay, callback, *args):
+    QtCore.QTimer.singleShot(delay * 1000, lambda: callback(*args))
+
+
+BUTTON_MAP = {
+    QtCore.Qt.MouseButton.LeftButton: 1,  # == MOUSE_BUTTON_LEFT
+    QtCore.Qt.MouseButton.RightButton: 2,  # == MOUSE_BUTTON_RIGHT
+    QtCore.Qt.MouseButton.MiddleButton: 3,  # == MOUSE_BUTTON_MIDDLE
+    QtCore.Qt.MouseButton.BackButton: 4,
+    QtCore.Qt.MouseButton.ForwardButton: 5,
+    QtCore.Qt.MouseButton.TaskButton: 6,
+    QtCore.Qt.MouseButton.ExtraButton4: 7,
+    QtCore.Qt.MouseButton.ExtraButton5: 8,
+}
+
+MODIFIERS_MAP = {
+    QtCore.Qt.ShiftModifier: "Shift",
+    QtCore.Qt.ControlModifier: "Control",
+    QtCore.Qt.AltModifier: "Alt",
+    QtCore.Qt.MetaModifier: "Meta",
+}
+
+KEY_MAP = {
+    int(QtCore.Qt.Key_Down): "ArrowDown",
+    int(QtCore.Qt.Key_Up): "ArrowUp",
+    int(QtCore.Qt.Key_Left): "ArrowLeft",
+    int(QtCore.Qt.Key_Right): "ArrowRight",
+    int(QtCore.Qt.Key_Backspace): "Backspace",
+    int(QtCore.Qt.Key_CapsLock): "CapsLock",
+    int(QtCore.Qt.Key_Delete): "Delete",
+    int(QtCore.Qt.Key_End): "End",
+    int(QtCore.Qt.Key_Enter): "Enter",
+    int(QtCore.Qt.Key_Escape): "Escape",
+    int(QtCore.Qt.Key_F1): "F1",
+    int(QtCore.Qt.Key_F2): "F2",
+    int(QtCore.Qt.Key_F3): "F3",
+    int(QtCore.Qt.Key_F4): "F4",
+    int(QtCore.Qt.Key_F5): "F5",
+    int(QtCore.Qt.Key_F6): "F6",
+    int(QtCore.Qt.Key_F7): "F7",
+    int(QtCore.Qt.Key_F8): "F8",
+    int(QtCore.Qt.Key_F9): "F9",
+    int(QtCore.Qt.Key_F10): "F10",
+    int(QtCore.Qt.Key_F11): "F11",
+    int(QtCore.Qt.Key_F12): "F12",
+    int(QtCore.Qt.Key_Home): "Home",
+    int(QtCore.Qt.Key_Insert): "Insert",
+    int(QtCore.Qt.Key_Alt): "Alt",
+    int(QtCore.Qt.Key_Control): "Control",
+    int(QtCore.Qt.Key_Shift): "Shift",
+    int(
+        QtCore.Qt.Key_Meta
+    ): "Meta",  # meta maps to control in QT on macOS, and vice-versa
+    int(QtCore.Qt.Key_NumLock): "NumLock",
+    int(QtCore.Qt.Key_PageDown): "PageDown",
+    int(QtCore.Qt.Key_PageUp): "Pageup",
+    int(QtCore.Qt.Key_Pause): "Pause",
+    int(QtCore.Qt.Key_ScrollLock): "ScrollLock",
+    int(QtCore.Qt.Key_Tab): "Tab",
+}
+
+
+class QAutoWgpuCanvas(WgpuCanvas):
+    def __init__(self, *args, **kwargs):
+        # When using Qt, there needs to be an
+        # application before any widget is created
+        global app
+        app = QtWidgets.QApplication.instance() or QtWidgets.QApplication([])
```

Comment:
A new application should only be created when `app` is still None.

Reply:
If you use this code, it should work also in `wgpu.gui.qt` ;) 

```py
global app
if QtWidgets.QApplication.instance() is None:
    QtWidgets.QApplication([])
if app is None:
    app = QtWidgets.QApplication.instance()
```

Reply:
I've added the check for `app is None`, good call.